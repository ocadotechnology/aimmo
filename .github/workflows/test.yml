name: Python style check
on:
  push:
    branches:
      - github_actions-1
jobs:
  build:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v1
    - name: Run a one-line script
      run: echo Hello, world!
    - name: Python style check
      uses: bulv1ne/python-style-check@v0.3
    - name: Run a multi-line script
      run: |
        echo Add other actions to build,
        echo test, and deploy your project.
