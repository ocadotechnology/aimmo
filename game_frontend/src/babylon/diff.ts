import isEqual from 'lodash.isequal'

export const ADD = 'A'
export const DELETE = 'D'
export const EDIT = 'E'

export class DiffItem {
  id: number
  value: any

  constructor (id: number, value: any) {
    this.id = id
    this.value = value
  }
}
export class DiffResult {
  addList: Array<DiffItem>
  deleteList: Array<DiffItem>
  editList: Array<DiffItem>

  constructor (toAdd: Array<DiffItem>, toDelete: Array<DiffItem>, toEdit: Array<DiffItem>) {
    this.addList = toAdd
    this.deleteList = toDelete
    this.editList = toEdit
  }
}

/**
 * Calculates the difference between to arrays.
 * @param previous - Previous state of an array
 * @param current  - New/Current state of an array
 *
 * @returns an object containing 3 lists for adding, removing, and editing items
 */
export function diff (previous: Array<any>, current: Array<any>): DiffResult {
  if (!previous.length) {
    return new DiffResult(convertToDiffItemsArray(current), [], [])
  }

  const arrayOfDifferences = getDifferingElements(previous, current)
  const editList = getItemsToEdit(arrayOfDifferences, current)
  const deleteList = getItemsToDelete(arrayOfDifferences, current)
  const remainingElements = processRemainingElements(previous, current)

  return new DiffResult(remainingElements, deleteList, editList)
}

function convertToDiffItemsArray (current: Array<any>): DiffItem[] {
  return current.map((element, index) => ({
    id: index,
    value: element
  }))
}

function getDifferingElements (previous: Array<any>, current: Array<any>): Array<Array<any>> {
  const differences = []
  for (const [index, element] of Object.entries(previous)) {
    if (!isEqual(element, current[index])) {
      differences.push([+index, element, current[index]])
    }
  }
  return differences
}

function getItemsToEdit (arrayOfDifferences: Array<Array<any>>, current: Array<any>): DiffItem[] {
  return arrayOfDifferences
    .filter(([id, previous, current]) => current !== undefined)
    .map(([id, previous, current]) => new DiffItem(id, current))
}

function getItemsToDelete (arrayOfDifferences: Array<Array<any>>, current: Array<any>): DiffItem[] {
  return arrayOfDifferences
    .filter(([id, previous, current]) => current === undefined)
    .map(([id, previous, current]) => new DiffItem(id, previous))
}

function processRemainingElements (previous: Array<any>, current: Array<any>): DiffItem[] {
  const remainingElements = []
  for (let index = previous.length; index < current.length; index++) {
    remainingElements.push(new DiffItem(index, current[index]))
  }
  return remainingElements
}
