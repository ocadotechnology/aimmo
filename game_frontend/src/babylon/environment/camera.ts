import { GameNode } from '../interfaces'
import * as BABYLON from 'babylonjs'
import { Environment } from './environment'

const ZOOM_LOWER_BOUND = 5
const ZOOM_UPPER_BOUND = 15
const ZOOM_COEFFICIENT = 20

const PANNING_SENSITITY = 6
const MAXIMUM_PANNING_SENSIBILITY = 130

export default class Camera implements GameNode {
    object: any
    frustum: number
    zoomFactor: number
    view: BABYLON.Vector2

    constructor (environment: Environment) {
      const camera = new BABYLON.ArcRotateCamera('camera1', 0, 0.785, 50, BABYLON.Vector3.Zero(), environment.scene)
      this.frustum = 7.5
      this.zoomFactor = 0
      this.object = camera

      camera.mode = BABYLON.Camera.ORTHOGRAPHIC_CAMERA
      camera.orthoTop = 5
      camera.orthoBottom = -5
      camera.orthoLeft = -5
      camera.orthoRight = 5
      camera.upperAlphaLimit = -(3 * Math.PI) / 4
      camera.lowerAlphaLimit = -(3 * Math.PI) / 4
      camera.upperBetaLimit = Math.PI / 4
      camera.lowerBetaLimit = Math.PI / 4
      camera.upperRadiusLimit = 50
      camera.lowerRadiusLimit = 50
      camera.panningInertia = 0
      camera.angularSensibilityX = 20
      camera.angularSensibilityY = 20
      camera.panningDistanceLimit = 20

      camera.setTarget(BABYLON.Vector3.Zero())

      camera.attachControl(environment.canvas, true, false, 0)

      this.updatePanningSensibility()
      this.computeCameraView(environment.canvas)

      this.addZoomListener(environment.scene, environment.canvas)
    }

    addZoomListener (scene: BABYLON.Scene, canvas: HTMLCanvasElement) {
      scene.onPrePointerObservable.add(pointerInfo => {
        const event = pointerInfo.event instanceof WheelEvent ? pointerInfo.event : undefined
        if (event) {
          let delta = 0
          if (event.deltaY) {
            delta = event.deltaY
          }

          if (delta) {
            this.zoomFactor += delta / ZOOM_COEFFICIENT

            if (this.zoomFactor + this.frustum >= ZOOM_UPPER_BOUND) {
              this.zoomFactor = ZOOM_UPPER_BOUND - this.frustum
            } else if (this.zoomFactor + this.frustum <= ZOOM_LOWER_BOUND) {
              this.zoomFactor = ZOOM_LOWER_BOUND - this.frustum
            }

            this.computeCameraView(canvas)
            this.updatePanningSensibility()
          }
        }
      }, BABYLON.PointerEventTypes.POINTERWHEEL, false)
    }

    updatePanningSensibility () {
      this.object.panningSensibility = -PANNING_SENSITITY *
            (this.frustum + this.zoomFactor) +
            MAXIMUM_PANNING_SENSIBILITY
    }

    computeCameraView (canvas: HTMLCanvasElement): void {
      this.view = new BABYLON.Vector2(canvas.width, canvas.height)
      this.view.normalize()

      this.object.orthoTop = (this.frustum + this.zoomFactor) * this.view.y
      this.object.orthoBottom = -(this.frustum + this.zoomFactor) * this.view.y
      this.object.orthoLeft = -(this.frustum + this.zoomFactor) * this.view.x
      this.object.orthoRight = (this.frustum + this.zoomFactor) * this.view.x
    }
}
