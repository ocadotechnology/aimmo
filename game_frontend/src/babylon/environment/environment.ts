import { Scene, Engine, TransformNode, Vector3 } from 'babylonjs'

export interface Environment {
  scene: Scene;
  engine: Engine;
  canvas?: HTMLCanvasElement;
  onTerrainNode: TransformNode;
}

export class StandardEnvironment implements Environment {
  scene: Scene;
  engine: Engine;
  canvas: HTMLCanvasElement;
  onTerrainNode: TransformNode;

  constructor (canvas: HTMLCanvasElement) {
    this.canvas = canvas
    this.engine = new Engine(this.canvas, true, {}, true)

    this.scene = new Scene(this.engine)

    this.onTerrainNode = new TransformNode('On Terrain', this.scene)
    this.onTerrainNode.position = new Vector3(0.5, 0, 0.5)
  }
}
