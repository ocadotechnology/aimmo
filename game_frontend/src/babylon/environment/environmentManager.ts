import Camera from './camera'
import Light from './light'
import Terrain from './terrain'
import { Environment } from './environment'

export default class EnvironmentManager {
    camera: Camera
    light: Light
    terrain: Terrain
    environment: Environment

    constructor (environment: Environment) {
      this.environment = environment

      this.camera = new Camera(environment)
      this.light = new Light(environment)
      this.terrain = new Terrain(environment)
    }

    resizeBabylonWindow = () => {
      this.environment.engine.resize()
      this.camera.computeCameraView(this.environment.canvas)
    }

    centerOn (mesh: any): void {
      this.camera.centerOn(mesh)
    }

    unCenter (mesh: any): void {
      this.camera.unCenter(mesh)
    }
}
