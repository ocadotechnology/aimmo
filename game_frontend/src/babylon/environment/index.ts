import { Environment } from './environment'

export default class SceneRenderer {
  environment: Environment

  constructor (environment: Environment) {
    this.environment = environment

    this.environment.engine.runRenderLoop(() => {
      this.environment.scene.render()
    })
    // this.showDebugLayer()
  }

  showDebugLayer (): void {
    setTimeout(() => {
      this.environment.scene.debugLayer.show(
        {
          overlay: true,
          showExplorer: true,
          showInspector: true,
          globalRoot: document.getElementById('root')
        }
      )
    }, 1000)
    setTimeout(() => {
      document.getElementById('scene-explorer-host').style.setProperty('z-index', '15000')
      document.getElementById('inspector-host').style.setProperty('z-index', '15000')
    }, 2000)
  }
}
