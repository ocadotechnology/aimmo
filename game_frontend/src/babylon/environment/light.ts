import { GameNode } from '../interfaces'
import { HemisphericLight, Vector3 } from 'babylonjs'
import { Environment } from './environment'

export default class Light implements GameNode {
    object: any;

    constructor (environment: Environment) {
      const light = new HemisphericLight('light1', new Vector3(0, 1, 0), environment.scene)
      this.object = light

      light.intensity = 1.45
    }
}
