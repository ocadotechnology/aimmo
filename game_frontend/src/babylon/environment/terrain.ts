import { GameNode } from '../interfaces'
import * as BABYLON from 'babylonjs'
import { Environment } from './environment'

export default class Terrain implements GameNode {
    object: any;

    constructor (environment: Environment) {
      const ground = BABYLON.Mesh.CreateTiledGround('terrain', -15, -15, 16, 16, { w: 31, h: 31 }, { w: 1, h: 1 }, environment.scene)
      this.object = ground

      const mat = new BABYLON.StandardMaterial('Terrain', environment.scene)
      mat.useReflectionOverAlpha = false
      mat.diffuseTexture = new BABYLON.Texture('/static/babylon/terrain/ground_future.jpg', environment.scene)
      mat.diffuseTexture.level = 1.2
      mat.specularColor = new BABYLON.Color3(0, 0, 0)
      mat.ambientColor = BABYLON.Color3.White()
      ground.material = mat
    }
}
