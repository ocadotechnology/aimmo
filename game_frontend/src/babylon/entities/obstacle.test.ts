/* eslint-env jest */
import { MockEnvironment } from '../environment/environment.test'
import Obstacle from './obstacle'
import { DiffResult } from '../diff'

let environment: MockEnvironment
let obstacle: Obstacle

beforeEach(() => {
  environment = new MockEnvironment()
  obstacle = new Obstacle()

  obstacle.setup(environment)

  addInitialObstacle('0', 10, 10)
})

function addInitialObstacle (id: string, posX: number, posY: number) {
  const addList = [{ id: id,
    value: {
      'location': { 'x': posX, 'y': posY },
      'width': 1,
      'height': 1,
      'type': 'wall',
      'orientation': 'north' }
  }]
  const diffResult = new DiffResult(addList, [], [])
  obstacle.onGameStateUpdate(diffResult)
}

describe('obstacle', () => {
  it('adds an Obstacles node', () => {
    const terrainNodeChildren = environment.onTerrainNode.getChildren()

    expect(terrainNodeChildren.length).toBe(1)
    expect(terrainNodeChildren[0].name).toBe('Obstacles')
  })

  it('adds an obstacle', () => {
    const obstacles = obstacle.obstacleNode.getChildren()

    expect(obstacles.length).toBe(1)
  })

  it('deletes an obstacle', () => {
    let obstacles = obstacle.obstacleNode.getChildren()

    expect(obstacles.length).toBe(1)

    const deleteList = [{ id: '0',
      value: {
        'location': { 'x': 10, 'y': 10 },
        'width': 1,
        'height': 1,
        'type': 'wall',
        'orientation': 'north' }
    }]
    const diffResult = new DiffResult([], deleteList, [])
    obstacle.onGameStateUpdate(diffResult)

    obstacles = obstacle.obstacleNode.getChildren()

    expect(obstacles.length).toBe(0)
  })

  it('edits an obstacle', () => {
    let obstacles = obstacle.obstacleNode.getChildren()

    expect(obstacles.length).toBe(1)
    expect(obstacles[0].position).toEqual({ x: 10, y: 0.5, z: 10 })

    const editList = [{ id: '0',
      value: {
        'location': { 'x': -7, 'y': 2 },
        'width': 1,
        'height': 1,
        'type': 'wall',
        'orientation': 'north' }
    }]
    const diffResult = new DiffResult([], [], editList)
    obstacle.onGameStateUpdate(diffResult)

    obstacles = obstacle.obstacleNode.getChildren()

    expect(obstacles.length).toBe(1)
    expect(obstacles[0].position).toEqual({ x: -7, y: 0.5, z: 2 })
  })

  it('adds, edits and delete obstacles on same update', () => {
    addInitialObstacle('1', 15, 15)

    let obstacles = obstacle.obstacleNode.getChildren()

    expect(obstacles.length).toBe(2)
    expect(obstacles[0].position).toEqual({ x: 10, y: 0.5, z: 10 })
    expect(obstacles[1].position).toEqual({ x: 15, y: 0.5, z: 15 })

    const addList = [{ id: '2',
      value: {
        'location': { 'x': 15, 'y': 15 },
        'width': 1,
        'height': 1,
        'type': 'wall',
        'orientation': 'north' }
    }]
    const editList = [{ id: '1',
      value: {
        'location': { 'x': 10, 'y': 10 },
        'width': 1,
        'height': 1,
        'type': 'wall',
        'orientation': 'north' }
    }]
    const deleteList = [{ id: '0',
      value: {
        'location': { 'x': 10, 'y': 10 },
        'width': 1,
        'height': 1,
        'type': 'wall',
        'orientation': 'north' }
    }]
    const diffResult = new DiffResult(addList, deleteList, editList)
    obstacle.onGameStateUpdate(diffResult)

    obstacles = obstacle.obstacleNode.getChildren()

    expect(obstacles.length).toBe(2)
    expect(obstacles[0].position).toEqual({ x: 10, y: 0.5, z: 10 })
    expect(obstacles[1].position).toEqual({ x: 15, y: 0.5, z: 15 })
  })
})
