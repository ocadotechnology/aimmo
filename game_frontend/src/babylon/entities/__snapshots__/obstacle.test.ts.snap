// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`obstacle adds an obstacle 1`] = `
Object {
  "name": "obstacle: 0",
  "position": e {
    "x": 10,
    "y": 0.5,
    "z": 10,
  },
}
`;

exports[`obstacle adds, edits and delete obstacles on same update 1`] = `
Object {
  "name": "obstacle: 0",
  "position": e {
    "x": 10,
    "y": 0.5,
    "z": 10,
  },
}
`;

exports[`obstacle adds, edits and delete obstacles on same update 2`] = `
Object {
  "name": "obstacle: 1",
  "position": e {
    "x": 15,
    "y": 0.5,
    "z": 15,
  },
}
`;

exports[`obstacle adds, edits and delete obstacles on same update 3`] = `
Object {
  "name": "obstacle: 1",
  "position": e {
    "x": 10,
    "y": 0.5,
    "z": 10,
  },
}
`;

exports[`obstacle adds, edits and delete obstacles on same update 4`] = `
Object {
  "name": "obstacle: 2",
  "position": e {
    "x": 15,
    "y": 0.5,
    "z": 15,
  },
}
`;

exports[`obstacle edits an obstacle 1`] = `
Object {
  "name": "obstacle: 0",
  "position": e {
    "x": 10,
    "y": 0.5,
    "z": 10,
  },
}
`;

exports[`obstacle edits an obstacle 2`] = `
Object {
  "name": "obstacle: 0",
  "position": e {
    "x": -7,
    "y": 0.5,
    "z": 2,
  },
}
`;
